integration-tests/tests/src/tests/sync/open-behavior.ts

```diff
@@ -0,0 +1,658 @@
+    // 1. Add data to server Realm from User
+    // 2. Open Realm again with User
+    // 3. Logout User and run timeout
+    // 4. It should not crash
+
+    const partitionValue = generatePartition();
+
+    const returningUserCredentials = await getRegisteredEmailPassCredentials(this.app);
+
+    {
+      // Add data to the realm with a different user
+      const user = await this.app.logIn(returningUserCredentials);
+      const config: Realm.Configuration = {
+        schema: [DogForSyncSchema],
+        sync: {
+          user,
+          partitionValue,
+          //@ts-expect-error TYPEBUG: cannot access const enum at runtime
+          _sessionStopPolicy: "immediately",
+        },
+      };
+
+      const realm = await Realm.open(config);
+
+      realm.write(() => {
+        realm.create(DogForSyncSchema.name, { _id: new ObjectId(), name: "Lola" });
+      });
+
+      await realm.syncSession?.uploadAllLocalChanges();
+
+      realm.close();
+      await user.logOut();
+    }
+
+    {
+      const user = await this.app.logIn(returningUserCredentials);
+      const config: Realm.Configuration = {
+        schema: [DogForSyncSchema],
+        sync: {
+          user,
+          partitionValue,
+          //@ts-expect-error TYPEBUG: cannot access const enum at runtime
+          _sessionStopPolicy: "immediately",
+          existingRealmFileBehavior: {
+            //@ts-expect-error TYPEBUG: cannot access const enum at runtime
+            type: "downloadBeforeOpen",
+            timeOut: 1000,
+            //@ts-expect-error TYPEBUG: cannot access const enum at runtime
+            timeOutBehavior: "openLocalRealm",
+          },
+        },
+      };
+      await Realm.open(config);
+
+      await user.logOut();
+
+      // Wait for the timeout to run.  This used to crash, since it opens a local realm with a logged out user.
+      await new Promise<void>((resolve) => {
+        sleep(1000);
+        resolve();
+      });
+    }
+  });
+  it("opening realm inside canceled promise", async function (this: AppContext) {
+    const user = await this.app.logIn(Realm.Credentials.anonymous());
+    const partitionValue = generatePartition();
+
+    const config: Realm.Configuration = {
+      schema: [DogForSyncSchema],
+      sync: {
+        user,
+        partitionValue,
+        //@ts-expect-error TYPEBUG: cannot access const enum at runtime
+        _sessionStopPolicy: "immediately",
+        //@ts-expect-error TYPEBUG: cannot access const enum at runtime
+        newRealmFileBehavior: { type: "downloadBeforeOpen" },
+      },
+    };
+
+    const openPromise = new Promise(() => {
+      const promise = Realm.open(config);
+      promise.cancel();
+      return promise;
+    });
+
+    openPromise
+      .then(() => {
+        expect(true).to.be.false("Realm was opened after being canceled.");
+      })
+      .catch((err) => {
+        expect(true).to.be.false("An error was thrown after open was canceled: " + err.message);
+      });
```

Comment:
You might as well throw here ... I see this pattern a few places below.

```suggestion
    const unexpectedPromise = openPromise
      .then(() => {
        throw new Error("Realm was opened after being canceled.");
      })
      .catch((err) => {
        throw new Error("An error was thrown after open was canceled: " + err.message);
      });
```

That being said, this will become unhandled promise rejections, so the result of this promise (`unexpectedPromise`) should probably be included in the `Promise.race` below.